### PR TITLE
some tweaks to handling 304 Not Modified responses

### LINKED ISSUE #3399: Unexpected token � in JSON at position 0.
**Description**

Hi,
Today I updated manually GitHub Desktop on Windows, from version 1.0.8 to 1.0.9.
After relaunching, I got the following error: 
> Unexpected token � in JSON at position 0.

It happens every time I launch GitHub Desktop.
Here is a screenshot: http://prntscr.com/hdetun
Note that it hasn't happened before.
Thanks.

**Version**

GitHub Desktop version: [1.0.9]
OS version: [Windows 10 Home version:1709]

**Steps to Reproduce**

1. Update manually by going to: Help->About GitHub Desktop -> Check for updates.
2.  Install update and wait for relaunch (the program relaunches by itself).

Expected behavior: No error.
Actual behavior: Error at program's launch

Reproduces how often: [100%]

**Logs**

```
2017-11-21T17:45:25.594Z - info: [ui] [AppStore] loading 1 repositories from store
2017-11-21T17:45:25.597Z - info: [ui] [AppStore] found account: LSmyrnaios (Lampros A. Smyrnaios)
2017-11-21T17:45:27.478Z - warn: [ui] Unable to deserialize JSON string to object [object Response]
SyntaxError: Unexpected token � in JSON at position 0
2017-11-21T17:45:27.479Z - warn: [ui] fetchPullRequests: failed for repository wheatvivo/datasources
SyntaxError: Unexpected token � in JSON at position 0
2017-11-21T17:45:27.480Z - warn: [ui] Error refreshing pull requests for 'datasources'
SyntaxError: Unexpected token � in JSON at position 0
2017-11-21T17:45:27.493Z - warn: [ui] Unable to deserialize JSON string to object [object Response]
TypeError: Failed to fetch
2017-11-21T17:45:27.495Z - warn: [ui] fetchMentionables: failed for wheatvivo/datasources
TypeError: Failed to fetch
2017-11-21T17:45:27.894Z - info: [ui] launching: 1.0.9 (Windows 10.0.16299)
```

### PR DESCRIPTION
**Pull request by shiftkey**
Fixes #3399 

 - [x] for `fetchAll`, if we've got a suspicious 304 let's log it and return an empty array instead
 - [x] make mentionables endpoint a bit more robust, unclear on root cause there
 - [x] the error handling in `deserialize<T>` doesn't stringify correctly - let's only look at some of the fields we're interested in when deserializing content

### CODE REVIEW

**Comment by shiftkey** [comment]
Looking at the original error log again, it seems like these are the two code paths that aren't being handled correctly:

```
2017-11-21T17:45:27.478Z - warn: [ui] Unable to deserialize JSON string to object [object Response]
SyntaxError: Unexpected token � in JSON at position 0
2017-11-21T17:45:27.479Z - warn: [ui] fetchPullRequests: failed for repository wheatvivo/datasources
SyntaxError: Unexpected token � in JSON at position 0
2017-11-21T17:45:27.480Z - warn: [ui] Error refreshing pull requests for 'datasources'
SyntaxError: Unexpected token � in JSON at position 0
```

This one can be traced back to `fetchAll` not expecting a 304. Although we don't set `If-None-Match` for this endpoint there's screenshots in #3399 that suggests the server can still trigger this.

I believe this is the error that's being displayed to the user in #3399, which will be handled better now.

```
2017-11-21T17:45:27.493Z - warn: [ui] Unable to deserialize JSON string to object [object Response]
TypeError: Failed to fetch
2017-11-21T17:45:27.495Z - warn: [ui] fetchMentionables: failed for wheatvivo/datasources
TypeError: Failed to fetch
```

This one is even more puzzling. Aside from storing an empty string instead of `null` in 37c3f886249000280bb2ccbff3b23bd3157c10b5 and seeing if a different status code ends up in `deserialize<T>` in 872e94401 I think this one needs more monitoring.

File: app/src/lib/api.ts [hunk 1]
```diff
@@ -506,18 +510,21 @@ export class API {
...
+
+      if (response.status === HttpStatusCode.NotFound) {
+        log.warn(`fetchMentionables: '${path}' returned a 404`)
         return null
       }
-      if (response.status === HttpStatusCode.NotFound) {
-        log.warn(`fetchAll: '${path}' returned a 404`)
+
+      if (response.status === HttpStatusCode.NotModified) {
+        log.warn(`fetchMentionables: '${path}' returned a 304`)
```
Comment (joshaber) [inline app/src/lib/api.ts]: Are we sure this deserves a warning? It seems like we'd expect a good amount of 304s from the mentionables endpoint.

Comment (shiftkey) [reply app/src/lib/api.ts]: Nope, 304s are pretty great status codes...

**Comment by shiftkey** [comment]
:apple:

**Pull request merged**